File: packages/native/android/src/main/java/expo/modules/jig/JigModule.kt
package expo.modules.jig

import expo.modules.jig.handlers.HandshakeHandler
import expo.modules.jig.handlers.ScreenshotHandler
import expo.modules.jig.middleware.DomainGuard
import expo.modules.jig.middleware.HandshakeGate
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class JigModule : Module() {
    private var server: JigWebSocketServer? = null

    override fun definition() = ModuleDefinition {
        Name("Jig")

        AsyncFunction("start") { config: Map<String, Any?> ->
            if (server != null) return@AsyncFunction

            val rnVersion = config["rnVersion"] as? String ?: "unknown"
            val expoVersion = config["expoVersion"] as? String
            val port = (config["port"] as? Double)?.toInt() ?: 4042

            val context = appContext.reactContext
                ?: throw Exception("React context not available")
            val androidAppInfo = context.packageManager
                .getApplicationInfo(context.packageName, 0)
            val appName = context.packageManager
                .getApplicationLabel(androidAppInfo).toString()

            val appInfo = JigAppInfo(
                appName, context.packageName, "android",
                rnVersion, expoVersion
            )

            val supportedDomains = emptyList<String>()
            val middlewares = listOf<JigMiddleware>(
                HandshakeGate(),
                DomainGuard(supportedDomains),
            )
            val handlers = listOf<JigHandler>(
                HandshakeHandler(),
                ScreenshotHandler { appContext.currentActivity },
            )
            val dispatcher = JigDispatcher(
                middlewares, handlers, supportedDomains
            )

            val newServer = JigWebSocketServer(port, dispatcher, appInfo)
            newServer.start()
            server = newServer
        }

        Function("stop") {
            server?.let { srv ->
                Thread {
                    try { srv.stop(1000) } catch (_: InterruptedException) {}
                }.start()
            }
            server = null
        }

        Function("isRunning") {
            server != null
        }
    }
}

File: packages/native/ios/JigModule.swift
// ios/JigModule.swift

import ExpoModulesCore

public class JigModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Jig")

        AsyncFunction("start") { (config: [String: Any]) in
            if JigBridge.isRunning() { return }

            let rnVersion = config["rnVersion"] as? String ?? "unknown"
            let expoVersion = config["expoVersion"] as? String
            let port = Int32(config["port"] as? Int ?? 4042)

            let appName = Bundle.main.object(
                forInfoDictionaryKey: "CFBundleName") as? String ?? "Unknown"
            let bundleId = Bundle.main.bundleIdentifier ?? "unknown"

            JigBridge.startServer(
                withName: appName,
                bundleId: bundleId,
                rnVersion: rnVersion,
                expoVersion: expoVersion,
                port: port
            )
        }

        Function("stop") {
            JigBridge.stopServer()
        }

        Function("isRunning") { () -> Bool in
            return JigBridge.isRunning()
        }
    }
}
